solidity/contracts/token/HypValue.sol

```diff
@@ -0,0 +1,166 @@
+     * @param _valuehook The address of the value transfer hook
+     * @param _interchainSecurityModule The address of the interchain security module
+     * @param _owner The owner of the contract
+     */
+    function initialize(
+        address _valuehook,
+        address _interchainSecurityModule,
+        address _owner
+    ) public initializer {
+        _MailboxClient_initialize(
+            _valuehook,
+            _interchainSecurityModule,
+            _owner
+        );
+    }
+
+    // ============ External Functions ============
+
+    /**
+     * @inheritdoc TokenRouter
+     * @dev use _hook with caution, make sure that this hook can handle msg.value transfer using the metadata.msgValue()
+     */
+    function transferRemote(
+        uint32 _destination,
+        bytes32 _recipient,
+        uint256 _amount,
+        bytes calldata _hookMetadata,
+        address _hook
+    ) external payable virtual override returns (bytes32 messageId) {
+        uint256 quote = _checkSufficientValue(_destination, _amount, _hook);
+
+        bytes memory hookMetadata = StandardHookMetadata.overrideMsgValue(
+            _hookMetadata,
+            _amount
+        );
+
+        return
+            _transferRemote(
+                _destination,
+                _recipient,
+                _amount,
+                _amount + quote,
+                hookMetadata,
+                _hook
+            );
+    }
+
+    /// @inheritdoc TokenRouter
+    function transferRemote(
+        uint32 _destination,
+        bytes32 _recipient,
+        uint256 _amount
+    ) external payable virtual override returns (bytes32 messageId) {
+        uint256 quote = _checkSufficientValue(
+            _destination,
+            _amount,
+            address(hook)
+        );
+        bytes memory hookMetadata = StandardHookMetadata.formatMetadata(
+            _amount,
+            destinationGas[_destination],
+            msg.sender,
+            ""
+        );
+
+        return
+            _transferRemote(
+                _destination,
+                _recipient,
+                _amount,
+                _amount + quote,
+                hookMetadata,
+                address(hook)
+            );
+    }
+
+    // ============ Internal Functions ============
+
+    /**
+     * @inheritdoc TokenRouter
+     * @dev No token metadata is needed for value transfers
+     */
+    function _transferFromSender(
+        uint256
+    ) internal pure override returns (bytes memory) {
+        return bytes(""); // no token metadata
+    }
+
+    /**
+     * @inheritdoc TokenRouter
+     * @dev Sends the value to the recipient
+     */
+    function _transferTo(
+        address _recipient,
+        uint256 _amount,
+        bytes calldata // no token metadata
+    ) internal virtual override {
+        Address.sendValue(payable(_recipient), _amount);
+    }
+
+    /**
+     * @inheritdoc TokenRouter
+     * @dev This contract doesn't hold value
```

Comment:
collateral balanceOf doesnt return what it holds, it returns what the user holds